File: super_grasp/src/super_grasp/grasps.py
#!/usr/bin/env python3.8

"""
    grasps: grasp point generation and selection from superquadric fits

    Input:
        sq: point cloud of the superquadric to fit superquadric to
        sq_poses: list of 3x3 rotation matrices of the superquadrics
        target_frame: frame id for the generated PoseStamped
        orientation: 'front' or 'top' for selecting grasp points based on object orientation
        gripper_width: width of the gripper for filtering grasp points
    
    Output:
        selectedGrasps: PoseStamped of the selected grasp point
        graspPoints: point cloud of all potential grasp points

    Developed by: Jayden Kanbour as part of undergraduate thesis for UNSW Computer Science and Engineering
    Date: 26th November 2025
    Email: jkanbour1@gmail.com
    UNSW Student Id: z5316799

"""

"""
Notes, 

Sample all ponts on the target SQ
    ✅ For front- get closest point on closest sq	

    ✅ For top- get highest point on highest sq

    ✅ Z-axis points to centre of sq
    
    X-axis points along the long extent
    Y-axis points along the short extent

"""

import rospy
import time
import numpy as np 
import open3d as o3d
from geometry_msgs.msg import PoseStamped
from scipy.spatial.transform import Rotation as R

class Grasps:
    def __init__(self, sq_list, target_frame, object_center, orientation=None, grasp_width=0.5, debug=False):

        self.debug = debug
        self.target_frame = target_frame
        self.orientation = orientation
        self.grasp_width = grasp_width
        self.object_center = object_center

        #generate and select best grasp
        if debug:
            self.graspPoints = o3d.geometry.PointCloud()
        
        time_start = time.time()

        self.primarySQ, self.grasp_point = self.SQFiltering(sq_list)
        time1 = time.time() - time_start


        self.selectedGrasps = self.generatePose(self.primarySQ, self.grasp_point, self.target_frame)
        time2 = time.time() - time_start - time1

        if self.debug:
            print(f"Grasps: Grasp points generated within {time1:.3f}s")
            print(f"Grasps: Grasp pose generated within {time2:.3f}s")

    def SQFiltering(self, sq_list):
        """
            Sort superquadrics by size and find valid grasp points.
            Returns the selected grasp point AND the corresponding OBB rotation matrix.
        """

        try:
            # Sort sq_list by their position relative to the camera origin (closest first)
            if self.orientation == 'front' or self.orientation is None:
                # Sort by Euclidean distance in the XY–Z plane (closest object to camera)
                # Sort both lists together based on 3D distance from origin
                sq_closest = sorted(sq_list, key=lambda x: np.linalg.norm(x.getCenter()[:3]))
                sq_closest = list(sq_closest)[0]

                sq_points = sq_closest.getSuperquadricMesh().points
                points = np.asarray(sq_points)
                distances_xz = np.sqrt(points[:, 0]**2 + points[:, 2]**2)
                closest_index = np.argmin(distances_xz)
                closest_point = points[closest_index]

                return sq_closest, closest_point

            elif self.orientation == 'top':
                # Sort by Y-axis (highest object first)

                # Sort both lists together based on superquadric Y-coordinate
                sq_highest = sorted(sq_list,
                                key=lambda x: x.getCenter()[1],
                                reverse=False)

                # Unzip back into separate lists
                sq_highest = sq_list(sq_highest)[0]
                sq_points = sq_closest.getSuperquadricMesh().points
                points = np.asarray(sq_points)
                # Find the index of the point with maximum y value
                highest_index = np.argmax(points[:, 1])

                # Get the highest point
                highest_point = points[highest_index]

                return sq_highest, highest_point
            
            return None
        
        except Exception as e:
            print(f"grasp [graspPointFiltering] Error: {e}")
            return None

    def generatePose(self, sq, grasp_point, frame_id):
        """
        Generate a PoseStamped by projecting a pose onto a point.
        """
        
        sq_center = sq.getCenter()
        bbox_extent = sq.getBBOXExtent()
        init_pose = sq.getSQPose()
        object_center = self.object_center
        
        try:
            # Create PoseStamped message
            pose_stamped = PoseStamped()
            pose_stamped.header.frame_id = frame_id
            pose_stamped.header.stamp = rospy.Time.now()
            
            # Set position from grasp point
            pose_stamped.pose.position.x = float(grasp_point[0])
            pose_stamped.pose.position.y = float(grasp_point[1])
            pose_stamped.pose.position.z = float(grasp_point[2])
            
            # Calculate orientation
            # a) Z-axis points towards object_center in x,y plane only
            # Create a target point at object_center's x,y but grasp_point's z
            target_point = np.array([object_center[0], object_center[1], grasp_point[2]])
            z_axis = target_point - np.array(grasp_point)
            z_axis = z_axis / np.linalg.norm(z_axis)  # Normalize
            
            # Find the shortest and longest extent axes
            extents = bbox_extent
            sorted_indices = np.argsort(extents)
            min_extent_idx = sorted_indices[0]  # Shortest
            max_extent_idx = sorted_indices[2]  # Longest
            
            # Get the axes in the superquadric's local frame
            shortest_axis_local = np.zeros(3)
            shortest_axis_local[min_extent_idx] = 1.0
            
            longest_axis_local = np.zeros(3)
            longest_axis_local[max_extent_idx] = 1.0
            
            # Transform to world frame using the pose rotation matrix
            # Extract rotation matrix from init_pose (assuming it's a 4x4 transformation matrix)
            rotation_matrix_sq = init_pose[:3, :3]
            
            shortest_axis_world = rotation_matrix_sq @ shortest_axis_local
            longest_axis_world = rotation_matrix_sq @ longest_axis_local
            
            # Project axes onto plane perpendicular to z_axis and assign to y and x
            # Y-axis should align with shortest extent
            y_axis = shortest_axis_world - np.dot(shortest_axis_world, z_axis) * z_axis
            y_axis_norm = np.linalg.norm(y_axis)
            
            # X-axis should align with longest extent
            x_axis = longest_axis_world - np.dot(longest_axis_world, z_axis) * z_axis
            x_axis_norm = np.linalg.norm(x_axis)
            
            # If one of the projections is too small, use cross product approach
            if y_axis_norm < 0.1 or x_axis_norm < 0.1:
                # Fallback: use the axis that has better projection
                if y_axis_norm > x_axis_norm:
                    y_axis = y_axis / y_axis_norm
                    x_axis = np.cross(y_axis, z_axis)
                    x_axis = x_axis / np.linalg.norm(x_axis)
                else:
                    x_axis = x_axis / x_axis_norm
                    y_axis = np.cross(z_axis, x_axis)
                    y_axis = y_axis / np.linalg.norm(y_axis)
            else:
                # Normalize both
                y_axis = y_axis / y_axis_norm
                x_axis = x_axis / x_axis_norm
                
                # Make sure they're orthogonal by adjusting x_axis
                x_axis = x_axis - np.dot(x_axis, y_axis) * y_axis
                x_axis = x_axis / np.linalg.norm(x_axis)
                
                # Ensure right-handed coordinate system
                if np.dot(np.cross(x_axis, y_axis), z_axis) < 0:
                    x_axis = -x_axis
            
            # Construct rotation matrix [x_axis, y_axis, z_axis]
            rotation_matrix = np.column_stack((x_axis, y_axis, z_axis))
            
            # Convert rotation matrix to quaternion
            scipy_rotation = R.from_matrix(rotation_matrix)
            quat = scipy_rotation.as_quat()  # Returns [x, y, z, w]
            
            # Set orientation
            pose_stamped.pose.orientation.x = float(quat[0])
            pose_stamped.pose.orientation.y = float(quat[1])
            pose_stamped.pose.orientation.z = float(quat[2])
            pose_stamped.pose.orientation.w = float(quat[3])
            
            if self.debug:
                print(f"grasp: Generated grasp pose at position: [{grasp_point[0]:.3f}, {grasp_point[1]:.3f}, {grasp_point[2]:.3f}]")
                print(f"grasp: Orientation (quaternion): [{quat[0]:.3f}, {quat[1]:.3f}, {quat[2]:.3f}, {quat[3]:.3f}]")
                print(f"grasp: Z-axis pointing to object center: [{z_axis[0]:.3f}, {z_axis[1]:.3f}, {z_axis[2]:.3f}]")
                print(f"grasp: Y-axis along shortest extent: [{y_axis[0]:.3f}, {y_axis[1]:.3f}, {y_axis[2]:.3f}]")
                print(f"grasp: X-axis along longest extent: [{x_axis[0]:.3f}, {x_axis[1]:.3f}, {x_axis[2]:.3f}]")
                print(f"grasp: Extents [x,y,z]: [{extents[0]:.3f}, {extents[1]:.3f}, {extents[2]:.3f}]")
                print(f"grasp: Shortest axis index: {min_extent_idx}, Longest axis index: {max_extent_idx}")
            
            return pose_stamped
        
        except Exception as e:      
            print(f"grasp [generatePose] Error: {e}")
            return None
        
    def getGraspPoints(self):
        return self.graspPoints
    
    def getSelectedGrasps(self):
        return self.selectedGrasps

File: super_grasp/src/super_grasp/grasps.cpp
#include "super_grasp/grasps.h"
#include <ros/ros.h>
#include <chrono>
#include <algorithm>

namespace super_grasp {

Grasps::Grasps(const std::vector<std::shared_ptr<open3d::geometry::PointCloud>>& sq,
               const std::vector<Eigen::Matrix3d>& sq_poses,
               const std::string& target_frame,
               const std::string& orientation,
               double grasp_width,
               bool debug)
    : target_frame_(target_frame),
      orientation_(orientation),
      grasp_width_(grasp_width),
      debug_(debug) {

    if (debug_) {
        grasp_points_ = std::make_shared<open3d::geometry::PointCloud>();
    }

    auto time_start = std::chrono::high_resolution_clock::now();

    // Filter and select best grasp point
    auto [point, pose] = graspPointFiltering(sq, sq_poses);
    primary_point_ = point;
    sq_pose_ = pose;

    auto time1 = std::chrono::high_resolution_clock::now();
    auto duration1 = std::chrono::duration_cast<std::chrono::milliseconds>(time1 - time_start);

    // Generate pose
    selected_grasps_ = generatePose(primary_point_, target_frame_, sq_pose_);

    auto time2 = std::chrono::high_resolution_clock::now();
    auto duration2 = std::chrono::duration_cast<std::chrono::milliseconds>(time2 - time1);

    if (debug_) {
        ROS_INFO("Grasps: Grasp points generated within %.3fs", duration1.count() / 1000.0);
        ROS_INFO("Grasps: Grasp pose generated within %.3fs", duration2.count() / 1000.0);
    }
}

std::pair<Eigen::Vector3d, Eigen::Matrix3d> Grasps::graspPointFiltering(
    const std::vector<std::shared_ptr<open3d::geometry::PointCloud>>& sq_list,
    const std::vector<Eigen::Matrix3d>& sq_poses) {

    try {
        if (sq_list.empty() || sq_poses.empty()) {
            ROS_WARN("Grasps: Empty superquadric list");
            return {Eigen::Vector3d::Zero(), Eigen::Matrix3d::Identity()};
        }

        // Create pairs of superquadrics and poses for sorting
        std::vector<std::pair<std::shared_ptr<open3d::geometry::PointCloud>, Eigen::Matrix3d>> sq_pairs;
        for (size_t i = 0; i < sq_list.size() && i < sq_poses.size(); ++i) {
            sq_pairs.push_back({sq_list[i], sq_poses[i]});
        }

        if (orientation_ == "front" || orientation_.empty()) {
            // Sort by Euclidean distance (closest to camera origin)
            std::sort(sq_pairs.begin(), sq_pairs.end(),
                [](const auto& a, const auto& b) {
                    auto obb_a = a.first->GetOrientedBoundingBox();
                    auto obb_b = b.first->GetOrientedBoundingBox();
                    return obb_a.center_.norm() < obb_b.center_.norm();
                });

            return {sq_pairs[0].first->GetCenter(), sq_pairs[0].second};

        } else if (orientation_ == "top") {
            // Sort by Y-axis (highest object first - lowest Y value in camera frame)
            std::sort(sq_pairs.begin(), sq_pairs.end(),
                [](const auto& a, const auto& b) {
                    auto obb_a = a.first->GetOrientedBoundingBox();
                    auto obb_b = b.first->GetOrientedBoundingBox();
                    return obb_a.center_(1) < obb_b.center_(1);
                });

            return {sq_pairs[0].first->GetCenter(), sq_pairs[0].second};
        }

        return {Eigen::Vector3d::Zero(), Eigen::Matrix3d::Identity()};

    } catch (const std::exception& e) {
        ROS_ERROR("grasp [graspPointFiltering] Error: %s", e.what());
        return {Eigen::Vector3d::Zero(), Eigen::Matrix3d::Identity()};
    }
}

Eigen::Quaterniond Grasps::rotationMatrixToQuaternion(const Eigen::Matrix3d& R) {
    // Convert rotation matrix to quaternion using Eigen
    Eigen::Quaterniond q(R);
    q.normalize();
    return q;
}

geometry_msgs::PoseStamped Grasps::generatePose(const Eigen::Vector3d& grasp_point,
                                                 const std::string& frame_id,
                                                 const Eigen::Matrix3d& sq_pose) {
    try {
        geometry_msgs::PoseStamped pose_stamped;
        pose_stamped.header.frame_id = frame_id;
        pose_stamped.header.stamp = ros::Time::now();

        // Set position
        pose_stamped.pose.position.x = grasp_point(0);
        pose_stamped.pose.position.y = grasp_point(1);
        pose_stamped.pose.position.z = grasp_point(2);

        // Convert rotation matrix to quaternion
        Eigen::Quaterniond scipy_rotation = rotationMatrixToQuaternion(sq_pose);

        if (orientation_ == "front" || orientation_.empty()) {
            // Apply 180° rotation about Y-axis
            Eigen::Quaterniond rotation_y_180(
                Eigen::AngleAxisd(M_PI, Eigen::Vector3d::UnitY()));
            scipy_rotation = scipy_rotation * rotation_y_180;

        } else { // "top" orientation
            // Apply 90° rotation about X-axis
            Eigen::Quaterniond rotation_x_90(
                Eigen::AngleAxisd(M_PI / 2.0, Eigen::Vector3d::UnitX()));
            scipy_rotation = scipy_rotation * rotation_x_90;
        }

        // Normalize and set orientation
        scipy_rotation.normalize();
        pose_stamped.pose.orientation.x = scipy_rotation.x();
        pose_stamped.pose.orientation.y = scipy_rotation.y();
        pose_stamped.pose.orientation.z = scipy_rotation.z();
        pose_stamped.pose.orientation.w = scipy_rotation.w();

        if (debug_) {
            ROS_INFO("grasp: Generated grasp pose at position: [%.3f, %.3f, %.3f]",
                     grasp_point(0), grasp_point(1), grasp_point(2));
            ROS_INFO("grasp: Orientation (quaternion): [%.3f, %.3f, %.3f, %.3f]",
                     scipy_rotation.x(), scipy_rotation.y(), 
                     scipy_rotation.z(), scipy_rotation.w());
        }

        return pose_stamped;

    } catch (const std::exception& e) {
        ROS_ERROR("grasp [generatePose] Error: %s", e.what());
        return geometry_msgs::PoseStamped();
    }
}

} // namespace super_grasp